# Processing the right output for each match

In [3]:
import pandas as pd
import numpy as np
from pathlib import Path
import random
current_dir = Path.cwd()
parent_dir = current_dir.parent.parent

In [4]:
innings = pd.read_pickle(parent_dir/"Resources"/"match_data.pkl")
final_data = pd.read_pickle(parent_dir/"Resources"/"final_data.pkl")

In [5]:
final_data

Player Status                                 \
                                Batting Team                                  
                                          P1   P2    P3   P4   P5   P6   P7   
Match Code Country   Over Ball                                                
211028     Australia 0    0             10.0  5.0   1.0  1.0  1.0  1.0  1.0   
                          1             10.0  5.0   1.0  1.0  1.0  1.0  1.0   
                          2             10.0  5.0   1.0  1.0  1.0  1.0  1.0   
                          3             10.0  5.0   1.0  1.0  1.0  1.0  1.0   
                          4             10.0  5.0   1.0  1.0  1.0  1.0  1.0   
...                                      ...  ...   ...  ...  ...  ...  ...   
1381457    Serbia    7    4             10.0  0.0   5.0  1.0  1.0  1.0  1.0   
                          5             10.0  0.0   5.0  1.0  1.0  1.0  1.0   
                     8    0              5.0  0.0  10.0  1.0  1.0  1.0  1.0   
                          1             10.0  0.0   5.0  1.0  1.0  1.0  1.0   
                          2              5.0  0.0  10.0  1.0  1.0  1.0  1.0   

                                               ...  \
                                               ...   
                                 P8   P9  P10  ...   
Match Code Country   Over Ball                 ...   
211028     Australia 0    0     1.0  1.0  1.0  ...   
                          1     1.0  1.0  1.0  ...   
                          2     1.0  1.0  1.0  ...   
                          3     1.0  1.0  1.0  ...   
                          4     1.0  1.0  1.0  ...   
...                             ...  ...  ...  ...   
1381457    Serbia    7    4     1.0  1.0  1.0  ...   
                          5     1.0  1.0  1.0  ...   
                     8    0     1.0  1.0  1.0  ...   
                          1     1.0  1.0  1.0  ...   
                          2     1.0  1.0  1.0  ...   

                                                                 Ground Data  \
                                                                           -   
                               Gahanga International Cricket Stadium, Rwanda   
Match Code Country   Over Ball                                                 
211028     Australia 0    0                                              0.0   
                          1                                              0.0   
                          2                                              0.0   
                          3                                              0.0   
                          4                                              0.0   
...                                                                      ...   
1381457    Serbia    7    4                                              0.0   
                          5                                              0.0   
                     8    0                                              0.0   
                          1                                              0.0   
                          2                                              0.0   

                                                                          \
                                                                           
                               Zahur Ahmed Chowdhury Stadium, Chattogram   
Match Code Country   Over Ball                                             
211028     Australia 0    0                                          0.0   
                          1                                          0.0   
                          2                                          0.0   
                          3                                          0.0   
                          4                                          0.0   
...                                                                  ...   
1381457    Serbia    7    4                                          0.0 

In [6]:
innings

batter         bowler  \
Match Code Country     Over Ball                                    
1381452    Austria     0    0     M Simpson-Parker  Ghulam Ahmadi   
                            1     M Simpson-Parker  Ghulam Ahmadi   
                            2     M Simpson-Parker  Ghulam Ahmadi   
                            3     M Simpson-Parker  Ghulam Ahmadi   
                            4     M Simpson-Parker  Ghulam Ahmadi   
...                                            ...            ...   
211048     New Zealand 19   1            JW Wilson     GD McGrath   
                            2            JW Wilson     GD McGrath   
                            3            DR Tuffey     GD McGrath   
                            4            DR Tuffey     GD McGrath   
                            5             KD Mills     GD McGrath   

                                    non_striker  batter runs  extra runs  \
Match Code Country     Over Ball                                           
1381452    Austria     0    0     Iqbal Hossain            0           0   
                            1     Iqbal Hossain            4           0   
                            2     Iqbal Hossain            0           0   
                            3     Iqbal Hossain            0           0   
                            4     Iqbal Hossain            1           0   
...                                         ...          ...         ...   
211048     New Zealand 19   1          KD Mills            2           0   
                            2          KD Mills            0           0   
                            3          KD Mills            4           0   
                            4          KD Mills            1           0   
                            5         DR Tuffey            0           0   

                                  total runs  score extras    wickets  out  \
Match Code Country     Over Ball                                             
1381452    Austria     0    0              0      0    NaN        NaN    0   
                            1              4      4    NaN        NaN    0   
                            2              0      4    NaN        NaN    0   
                            3              0      4    NaN        NaN    0   
                            4              1      5    NaN        NaN    0   
...                                      ...    ...    ...        ...  ...   
211048     New Zealand 19   1              2    165    NaN        NaN    0   
                            2              0    165    NaN  JW Wilson    1   
                            3              4    169    NaN        NaN    0   
                            4              1    170    NaN        NaN    0   
                            5              0    170    NaN   KD Mills    1   

                                  Total Wickets non_boundary review  \
Match Code Country     Over Ball                                      
1381452    Austria     0    0                 0          NaN    NaN   
                            1                 0          NaN    NaN   
                            2                 0          NaN    NaN   
                            3                 0          NaN    NaN   
                            4                 0          NaN    NaN   
...                                         ...          ...    ...   
211048     New Zealand 19   1                 8          NaN    NaN   
                            2                 9          NaN    NaN   
                            3                 9          NaN    NaN   
                            4                 9          NaN    NaN   
                            5                10          NaN    NaN   

                                 replacements  
Match Code Country     Over Ball               
1381452    Austria     0    0             NaN  
                            1             NaN  
                          

In [7]:
innings_grouped = innings['score'].groupby(['Match Code', 'Country'])

In [8]:
output_scores = innings_grouped.apply(max).to_frame()
output_scores.columns = pd.MultiIndex.from_product([['Final Output'], ['Output Score'], ['Score']])
output_scores

Final Output
                        Output Score
                               Score
Match Code Country                  
1381452    Austria               121
           Germany               123
1381451    Austria               114
           Germany               115
1381456    Croatia                83
...                              ...
222678     South Africa          133
211028     Australia              79
           England               179
211048     Australia             214
           New Zealand           170

[3525 rows x 1 columns]

In [9]:
final_data.reset_index(["Over", "Ball"])

Over Ball Player Status                                 \
                                Batting Team                                  
                                          P1   P2    P3   P4   P5   P6   P7   
Match Code Country                                                            
211028     Australia    0    0          10.0  5.0   1.0  1.0  1.0  1.0  1.0   
           Australia    0    1          10.0  5.0   1.0  1.0  1.0  1.0  1.0   
           Australia    0    2          10.0  5.0   1.0  1.0  1.0  1.0  1.0   
           Australia    0    3          10.0  5.0   1.0  1.0  1.0  1.0  1.0   
           Australia    0    4          10.0  5.0   1.0  1.0  1.0  1.0  1.0   
...                   ...  ...           ...  ...   ...  ...  ...  ...  ...   
1381457    Serbia       7    4          10.0  0.0   5.0  1.0  1.0  1.0  1.0   
           Serbia       7    5          10.0  0.0   5.0  1.0  1.0  1.0  1.0   
           Serbia       8    0           5.0  0.0  10.0  1.0  1.0  1.0  1.0   
           Serbia       8    1          10.0  0.0   5.0  1.0  1.0  1.0  1.0   
           Serbia       8    2           5.0  0.0  10.0  1.0  1.0  1.0  1.0   

                           ...                                   Ground Data  \
                           ...                                             -   
                       P8  ... Gahanga International Cricket Stadium, Rwanda   
Match Code Country         ...                                                 
211028     Australia  1.0  ...                                           0.0   
           Australia  1.0  ...                                           0.0   
           Australia  1.0  ...                                           0.0   
           Australia  1.0  ...                                           0.0   
           Australia  1.0  ...                                           0.0   
...                   ...  ...                                           ...   
1381457    Serbia     1.0  ...                                           0.0   
           Serbia     1.0  ...                                           0.0   
           Serbia     1.0  ...                                           0.0   
           Serbia     1.0  ...                                           0.0   
           Serbia     1.0  ...                                           0.0   

                                                                \
                                                                 
                     Zahur Ahmed Chowdhury Stadium, Chattogram   
Match Code Country                                               
211028     Australia                                       0.0   
           Australia                                       0.0   
           Australia                                       0.0   
           Australia                                       0.0   
           Australia                                       0.0   
...                                                        ...   
1381457    Serbia                                          0.0   
           Serbia                                          0.0   
           Serbia                                          0.0   
           Serbia                                          0.0   
           Serbia                                          0.0   

                                                   \
                                                    
                     St Albans Club, Buenos Aires   
Match Code Country                                  
211028     Australia                          0.0   
           Australia                          0.0   
           Australia                          0.0   
           Australia                          0.0   
           Australia                          0.0   
...                                           ...   
1381457    Serbia                             0.0   
           Serbia                             0.0   
       

In [10]:
final_data_with_output = final_data.merge(output_scores, left_index = True, right_index = True, how = 'left')

final_data_with_output

Player Status                                 \
                                Batting Team                                  
                                          P1   P2    P3   P4   P5   P6   P7   
Match Code Country   Over Ball                                                
211028     Australia 0    0             10.0  5.0   1.0  1.0  1.0  1.0  1.0   
                          1             10.0  5.0   1.0  1.0  1.0  1.0  1.0   
                          2             10.0  5.0   1.0  1.0  1.0  1.0  1.0   
                          3             10.0  5.0   1.0  1.0  1.0  1.0  1.0   
                          4             10.0  5.0   1.0  1.0  1.0  1.0  1.0   
...                                      ...  ...   ...  ...  ...  ...  ...   
1381457    Serbia    7    4             10.0  0.0   5.0  1.0  1.0  1.0  1.0   
                          5             10.0  0.0   5.0  1.0  1.0  1.0  1.0   
                     8    0              5.0  0.0  10.0  1.0  1.0  1.0  1.0   
                          1             10.0  0.0   5.0  1.0  1.0  1.0  1.0   
                          2              5.0  0.0  10.0  1.0  1.0  1.0  1.0   

                                               ...  \
                                               ...   
                                 P8   P9  P10  ...   
Match Code Country   Over Ball                 ...   
211028     Australia 0    0     1.0  1.0  1.0  ...   
                          1     1.0  1.0  1.0  ...   
                          2     1.0  1.0  1.0  ...   
                          3     1.0  1.0  1.0  ...   
                          4     1.0  1.0  1.0  ...   
...                             ...  ...  ...  ...   
1381457    Serbia    7    4     1.0  1.0  1.0  ...   
                          5     1.0  1.0  1.0  ...   
                     8    0     1.0  1.0  1.0  ...   
                          1     1.0  1.0  1.0  ...   
                          2     1.0  1.0  1.0  ...   

                                                             Ground Data  \
                                                                       -   
                               Zahur Ahmed Chowdhury Stadium, Chattogram   
Match Code Country   Over Ball                                             
211028     Australia 0    0                                          0.0   
                          1                                          0.0   
                          2                                          0.0   
                          3                                          0.0   
                          4                                          0.0   
...                                                                  ...   
1381457    Serbia    7    4                                          0.0   
                          5                                          0.0   
                     8    0                                          0.0   
                          1                                          0.0   
                          2                                          0.0   

                                                             \
                                                              
                               St Albans Club, Buenos Aires   
Match Code Country   Over Ball                                
211028     Australia 0    0                             0.0   
                          1                             0.0   
                          2                             0.0   
                          3                             0.0   
                          4                             0.0   
...                                                     ...   
1381457    Serbia    7    4                             0.0   
                          5                             0.0   
                     8    0                             0.0   
                          1                             0.0   
           

## Dividing the data

#### In order to divide the data, I will be working with its index and select random matches from it :)}

In [11]:
final_data_with_output.index.levshape
df = final_data_with_output
df.index.levshape

(1231, 91, 20, 12)

This Shows there are a total of 1231 Matches. It will hence be a good idea to divide it into the following Data Set:

| Training Set | Dev Set | Test Set |
| --- | --- | --- |
| 1100 | 100 | 31 |

In [12]:
match_codes = df.index.get_level_values(0).unique().to_list()
random.shuffle(match_codes)
match_codes

[1310170,
 387564,
 1273148,
 533279,
 1277975,
 521217,
 430884,
 1268192,
 1320980,
 1317486,
 1321277,
 1343740,
 1223954,
 1184258,
 1263474,
 412677,
 1074966,
 1273134,
 1298167,
 350476,
 1188623,
 540174,
 1133817,
 1310172,
 356004,
 1349131,
 1199515,
 287867,
 1286670,
 1287775,
 1142506,
 1359787,
 1233961,
 635660,
 1198236,
 1283027,
 1321475,
 1187668,
 1282278,
 298795,
 1273143,
 1320969,
 802327,
 1349385,
 533273,
 514023,
 1349134,
 1267307,
 1349128,
 1307478,
 915785,
 884349,
 1197525,
 1120291,
 1343793,
 1279379,
 1310177,
 954735,
 1199549,
 251488,
 1348325,
 1122286,
 1197524,
 817205,
 1199506,
 520595,
 1275276,
 592268,
 1130746,
 1001353,
 722337,
 1273273,
 356013,
 951323,
 1157375,
 354456,
 647247,
 1286672,
 249227,
 1197400,
 1278680,
 1075508,
 1198245,
 1298161,
 1273729,
 951305,
 1263710,
 351694,
 1115807,
 1197403,
 1273726,
 963697,
 361531,
 1343750,
 1142503,
 1133819,
 1305503,
 1270838,
 412701,
 1072319,
 1303314,
 1199539,
 963699,
 53

In [13]:
test_set_codes = match_codes[:31]
dev_set_codes = match_codes[31:131]
training_set_codes = match_codes[131:]

print(len(test_set_codes), len(dev_set_codes), len(training_set_codes))


31 100 1100


In [14]:
test_list = []
dev_list = []
training_list = []

for code in test_set_codes:
    test_list.append(df.loc[code])

for code in dev_set_codes:
    dev_list.append(df.loc[code])
    
for code in training_set_codes:
    training_list.append(df.loc[code])

In [15]:
test_data = pd.concat(test_list, keys = test_set_codes)
dev_data = pd.concat(dev_list, keys = dev_set_codes)
training_data = pd.concat(training_list, keys = training_set_codes)

In [16]:
test_data

Player Status                                   \
                             Batting Team                                    
                                       P1    P2   P3    P4    P5   P6   P7   
        Country   Over Ball                                                  
1310170 Gibraltar 0    0             10.0   5.0  1.0   1.0   1.0  1.0  1.0   
                       1              5.0  10.0  1.0   1.0   1.0  1.0  1.0   
                       2              5.0  10.0  1.0   1.0   1.0  1.0  1.0   
                       3              5.0  10.0  1.0   1.0   1.0  1.0  1.0   
                       4              5.0  10.0  1.0   1.0   1.0  1.0  1.0   
...                                   ...   ...  ...   ...   ...  ...  ...   
1142506 Scotland  19   1              0.0   0.0  5.0  10.0   1.0  1.0  1.0   
                       2              0.0   0.0  0.0   5.0   1.0  1.0  1.0   
                       3              0.0   0.0  0.0  10.0   5.0  1.0  1.0   
                       4              0.0   0.0  0.0  10.0   5.0  1.0  1.0   
                       5              0.0   0.0  0.0   5.0  10.0  1.0  1.0   

                                            ...  \
                                            ...   
                              P8   P9  P10  ...   
        Country   Over Ball                 ...   
1310170 Gibraltar 0    0     1.0  1.0  1.0  ...   
                       1     1.0  1.0  1.0  ...   
                       2     1.0  1.0  1.0  ...   
                       3     1.0  1.0  1.0  ...   
                       4     1.0  1.0  1.0  ...   
...                          ...  ...  ...  ...   
1142506 Scotland  19   1     1.0  1.0  1.0  ...   
                       2     1.0  1.0  1.0  ...   
                       3     1.0  1.0  1.0  ...   
                       4     1.0  1.0  1.0  ...   
                       5     1.0  1.0  1.0  ...   

                                                          Ground Data  \
                                                                    -   
                            Zahur Ahmed Chowdhury Stadium, Chattogram   
        Country   Over Ball                                             
1310170 Gibraltar 0    0                                          0.0   
                       1                                          0.0   
                       2                                          0.0   
                       3                                          0.0   
                       4                                          0.0   
...                                                               ...   
1142506 Scotland  19   1                                          0.0   
                       2                                          0.0   
                       3                                          0.0   
                       4                                          0.0   
                       5                                          0.0   

                                                          \
                                                           
                            St Albans Club, Buenos Aires   
        Country   Over Ball                                
1310170 Gibraltar 0    0                             0.0   
                       1                             0.0   
                       2                             0.0   
                       3                             0.0   
                       4                             0.0   
...                                                  ...   
1142506 Scotland  19   1                             0.0   
                       2                             0.0   
                       3                             0.0   
                       4                             0.0   
                       5                             0.0   

                                                                         \
                   

In [17]:
dev_data

Player Status                                 \
                               Batting Team                                  
                                         P1   P2   P3   P4    P5   P6   P7   
        Country     Over Ball                                                
1359787 Argentina   0    0             10.0  5.0  1.0  1.0   1.0  1.0  1.0   
                         1             10.0  5.0  1.0  1.0   1.0  1.0  1.0   
                         2             10.0  5.0  1.0  1.0   1.0  1.0  1.0   
                         3             10.0  5.0  1.0  1.0   1.0  1.0  1.0   
                         4             10.0  5.0  1.0  1.0   1.0  1.0  1.0   
...                                     ...  ...  ...  ...   ...  ...  ...   
1273746 West Indies 19   1              0.0  0.0  0.0  0.0  10.0  0.0  0.0   
                         2              0.0  0.0  0.0  0.0  10.0  0.0  0.0   
                         3              0.0  0.0  0.0  0.0  10.0  0.0  0.0   
                         4              0.0  0.0  0.0  0.0  10.0  0.0  0.0   
                         5              0.0  0.0  0.0  0.0  10.0  0.0  0.0   

                                              ...  \
                                              ...   
                                P8   P9  P10  ...   
        Country     Over Ball                 ...   
1359787 Argentina   0    0     1.0  1.0  1.0  ...   
                         1     1.0  1.0  1.0  ...   
                         2     1.0  1.0  1.0  ...   
                         3     1.0  1.0  1.0  ...   
                         4     1.0  1.0  1.0  ...   
...                            ...  ...  ...  ...   
1273746 West Indies 19   1     0.0  0.0  5.0  ...   
                         2     0.0  0.0  5.0  ...   
                         3     0.0  0.0  5.0  ...   
                         4     0.0  0.0  5.0  ...   
                         5     0.0  0.0  5.0  ...   

                                                            Ground Data  \
                                                                      -   
                              Zahur Ahmed Chowdhury Stadium, Chattogram   
        Country     Over Ball                                             
1359787 Argentina   0    0                                          0.0   
                         1                                          0.0   
                         2                                          0.0   
                         3                                          0.0   
                         4                                          0.0   
...                                                                 ...   
1273746 West Indies 19   1                                          0.0   
                         2                                          0.0   
                         3                                          0.0   
                         4                                          0.0   
                         5                                          0.0   

                                                            \
                                                             
                              St Albans Club, Buenos Aires   
        Country     Over Ball                                
1359787 Argentina   0    0                             0.0   
                         1                             0.0   
                         2                             0.0   
                         3                             0.0   
                         4                             0.0   
...                                                    ...   
1273746 West Indies 19   1                             0.0   
                         2                             0.0   
                         3                             0.0   
                         4                             0.0   
                         5                             0.0   

      

In [18]:
training_data

Player Status                              \
                                Batting Team                               
                                          P1    P2    P3   P4   P5    P6   
        Country      Over Ball                                             
1144161 Pakistan     0    0             10.0   5.0   1.0  1.0  1.0   1.0   
                          1             10.0   5.0   1.0  1.0  1.0   1.0   
                          2              0.0   5.0   1.0  1.0  1.0   1.0   
                          3              0.0   5.0  10.0  1.0  1.0   1.0   
                          4              0.0  10.0   5.0  1.0  1.0   1.0   
...                                      ...   ...   ...  ...  ...   ...   
356014  South Africa 19   1              0.0   0.0   0.0  0.0  0.0   5.0   
                          2              0.0   0.0   0.0  0.0  0.0  10.0   
                          3              0.0   0.0   0.0  0.0  0.0  10.0   
                          4              0.0   0.0   0.0  0.0  0.0   5.0   
                          5              0.0   0.0   0.0  0.0  0.0   5.0   

                                                     ...  \
                                                     ...   
                                  P7   P8   P9  P10  ...   
        Country      Over Ball                       ...   
1144161 Pakistan     0    0      1.0  1.0  1.0  1.0  ...   
                          1      1.0  1.0  1.0  1.0  ...   
                          2      1.0  1.0  1.0  1.0  ...   
                          3      1.0  1.0  1.0  1.0  ...   
                          4      1.0  1.0  1.0  1.0  ...   
...                              ...  ...  ...  ...  ...   
356014  South Africa 19   1      1.0  1.0  1.0  1.0  ...   
                          2      5.0  1.0  1.0  1.0  ...   
                          3      5.0  1.0  1.0  1.0  ...   
                          4     10.0  1.0  1.0  1.0  ...   
                          5     10.0  1.0  1.0  1.0  ...   

                                                             Ground Data  \
                                                                       -   
                               Zahur Ahmed Chowdhury Stadium, Chattogram   
        Country      Over Ball                                             
1144161 Pakistan     0    0                                          0.0   
                          1                                          0.0   
                          2                                          0.0   
                          3                                          0.0   
                          4                                          0.0   
...                                                                  ...   
356014  South Africa 19   1                                          0.0   
                          2                                          0.0   
                          3                                          0.0   
                          4                                          0.0   
                          5                                          0.0   

                                                             \
                                                              
                               St Albans Club, Buenos Aires   
        Country      Over Ball                                
1144161 Pakistan     0    0                             0.0   
                          1                             0.0   
                          2                             0.0   
                          3                             0.0   
                          4                             0.0   
...                                                     ...   
356014  South Africa 19   1                             0.0   
                          2                             0.0   
                          3                             0.0   
                          4  

# Saving the data

### We will be saving the data in a sep folder

In [19]:
training_data.to_pickle(parent_dir/'Resources'/'FinalData'/'training_data.pkl')
test_data.to_pickle(parent_dir/'Resources'/'FinalData'/'test_data.pkl')
dev_data.to_pickle(parent_dir/'Resources'/'FinalData'/'dev_data.pkl')